File: windows_player/discovery.py
"""UDP discovery responder (§7).

Players listen on UDP 8772. A controller/broker broadcasts a `discover`
envelope; the player unicasts back an `announce` envelope (HMAC-signed like any
other message, §3) carrying device_id/name/ip and a broker_hint. Control still
flows over the broker WS; UDP is only for list refresh / IP backfill / fallback.

Runs on its own daemon thread with a blocking socket — independent of the
asyncio WS loop.
"""

from __future__ import annotations

import json
import socket
import threading
from typing import Any, Callable, Dict, Optional

import envelope
import auth as auth_mod


class DiscoveryResponder:
    def __init__(self, *, psk: str, device_id: str, device_name: str,
                 ip: str, broker_hint: str, port: int = 8772,
                 replay: Optional[envelope.ReplayCache] = None,
                 auth_mode: str = "open", topology: str = "dedicated",
                 key_mode: str = envelope.KEY_MODE_GLOBAL,
                 device_key: Optional[bytes] = None,
                 verify_keys: Optional[dict] = None):
        self.psk = psk
        self.device_id = device_id
        self.device_name = device_name
        self.ip = ip
        self.broker_hint = broker_hint
        self.port = port
        self.replay = replay or envelope.ReplayCache()
        # §13/§14: declare the coordinator's auth_mode + topology so probing
        # endpoints adapt (signing) and diagnose (role) without connecting.
        self.auth_mode = auth_mod.normalize_mode(auth_mode)
        self.topology = topology
        # §17: key_mode for signing our announce / verifying inbound discover.
        self.key_mode = auth_mod.normalize_key_mode(key_mode)
        self.device_key = device_key
        self.verify_keys = dict(verify_keys or {})
        self._sock: Optional[socket.socket] = None
        self._thread: Optional[threading.Thread] = None
        self._stop = threading.Event()

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, name="udp-discovery",
                                        daemon=True)
        self._thread.start()

    def _make_announce(self) -> bytes:
        # §13: sign the announce only when our mode calls for it (open → "").
        frm = f"player:{self.device_id}"
        has_material = auth_mod.has_usable_psk(self.psk) or (
            self.key_mode == envelope.KEY_MODE_DERIVED
            and self.device_key is not None)
        sign = auth_mod.should_sign(self.auth_mode, has_material)
        payload = {
            "device_id": self.device_id,
            "device_name": self.device_name,
            "ip": self.ip,
            "broker_hint": self.broker_hint,
            "auth_mode": self.auth_mode,
            "topology": self.topology,
            "key_mode": self.key_mode,
        }
        env = envelope.build_envelope(
            self.psk, "announce", frm, "all", payload,
            sign_frame=sign, key_mode=self.key_mode, device_key=self.device_key)
        return json.dumps(env, ensure_ascii=False).encode("utf-8")

    def _run(self) -> None:
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            try:
                s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            except Exception:
                pass
            s.bind(("", self.port))
            s.settimeout(1.0)
            self._sock = s
        except Exception:
            return
        while not self._stop.is_set():
            try:
                data, addr = self._sock.recvfrom(8192)
            except socket.timeout:
                continue
            except Exception:
                break
            self._handle(data, addr)

    def _handle(self, data: bytes, addr) -> None:
        try:
            env = json.loads(data.decode("utf-8"))
        except Exception:
            return
        if env.get("type") != "discover":
            return
        resolver = None
        verify_km = self.key_mode
        if self.key_mode == envelope.KEY_MODE_DERIVED and \
                not auth_mod.has_usable_psk(self.psk):
            keys = self.verify_keys
            resolver = lambda frm: keys.get(frm)  # noqa: E731
        ok, _ = envelope.verify(self.psk, env, replay=self.replay,
                                first_connect=True, auth_mode=self.auth_mode,
                                key_mode=verify_km, key_resolver=resolver)
        if not ok:
            return
        try:
            self._sock.sendto(self._make_announce(), addr)  # unicast reply
        except Exception:
            pass

    def update_ip(self, ip: str) -> None:
        self.ip = ip

    def stop(self) -> None:
        self._stop.set()
        try:
            if self._sock:
                self._sock.close()
        except Exception:
            pass

File: android_apps/player/app/src/main/kotlin/com/jieoz/lanmediawall/player/net/Discovery.kt
package com.jieoz.lanmediawall.player.net

import android.util.Log
import java.net.DatagramPacket
import java.net.DatagramSocket
import kotlin.concurrent.thread

/**
 * UDP discovery responder — protocol_spec §7. The Android analogue of
 * windows_player/discovery.py.
 *
 * Listens on UDP [port] (8772). When a controller/broker broadcasts a
 * `discover` envelope, we unicast back an `announce` envelope (HMAC-signed like
 * any other message, §3) carrying device_id/name/ip + broker_hint. Control
 * still flows over the broker WS; UDP is only for list refresh / IP backfill.
 *
 * Runs on its own daemon thread with a blocking socket — independent of the WS
 * client.
 */
class Discovery(
    private val psk: String,
    private val deviceId: String,
    @Volatile private var deviceName: String,
    @Volatile private var ip: String,
    @Volatile private var brokerHint: String,
    private val port: Int = 8772,
    /**
     * §14/§13/§17.3: the topology this device advertises, and the auth/key mode
     * the coordinator declares. **Null `topology` = today's broker-client path**
     * — the announce payload remains unchanged (device_id/name/ip/broker_hint),
     * while its signature follows the active auth/key mode. When
     * set (e.g. "p2p" — this device IS the coordinator, §14.3), the announce
     * additionally carries `topology`/`auth_mode`/`key_mode` and is signed under
     * that auth/key mode, matching windows_player/discovery.py::DiscoveryResponder.
     */
    @Volatile private var topology: String? = null,
    private val authMode: AuthMode = AuthMode.OPTIONAL,
    private val keyMode: KeyMode = KeyMode.GLOBAL,
    private val deviceKey: ByteArray? = null,
) {
    @Volatile private var socket: DatagramSocket? = null
    @Volatile private var running = false
    private val replay = ReplayCache()

    fun start() {
        if (running) return
        running = true
        thread(name = "udp-discovery", isDaemon = true) { run() }
    }

    private fun run() {
        val sock = try {
            // QZX_C1/YunOS 4.4 can throw IllegalArgumentException("port=-1")
            // from the DatagramSocket(null)+InetSocketAddress(port) path even
            // when [port] is 8772. The one-arg constructor is the older Android
            // code path and binds INADDR_ANY:port directly, matching what we need.
            DatagramSocket(port).apply {
                reuseAddress = true
                broadcast = true
                soTimeout = 1000
            }
        } catch (e: Exception) {
            Log.w(TAG, "UDP discovery bind failed on $port: ${e.javaClass.simpleName}: ${e.message}")
            running = false
            return
        }
        socket = sock
        Log.i(TAG, "UDP discovery responder listening on $port topology=${topology ?: "broker"} authMode=${authMode.wire} keyMode=${keyMode.wire}")
        val buf = ByteArray(8192)
        while (running) {
            val packet = DatagramPacket(buf, buf.size)
            try {
                sock.receive(packet)
            } catch (e: java.net.SocketTimeoutException) {
                continue
            } catch (e: Exception) {
                if (!running) break else continue
            }
            // An exception on an Android background thread terminates the whole
            // process. Discovery is advisory, so malformed packets/configuration
            // must be contained here rather than taking the kiosk down.
            try {
                handle(packet, sock)
            } catch (e: Exception) {
                Log.e(TAG, "UDP discovery packet failed: ${e.javaClass.simpleName}: ${e.message}")
            }
        }
        try { sock.close() } catch (_: Exception) {}
    }

    private fun handle(packet: DatagramPacket, sock: DatagramSocket) {
        val text = String(packet.data, packet.offset, packet.length, Charsets.UTF_8)
        val result = Envelope.verify(
            psk,
            text,
            replay = replay,
            firstConnect = true,
            authMode = authMode,
            keyMode = keyMode,
            verifyKeyFor = ::verifyKeyFor,
        )
        if (!result.ok || result.parsed == null) {
            val peek = Envelope.peekTypeFrom(text)
            Log.w(TAG, "DROP discovery inbound: reason=${result.reason} type=${peek?.first ?: "?"} from=${peek?.second ?: "?"} sigLen=${peek?.third ?: -1} authMode=${authMode.wire} keyMode=${keyMode.wire}")
            return
        }
        if (result.parsed.type != "discover") return
        val reply = makeAnnounce()
        if (reply == null) {
            // REQUIRED without usable key material is a broken configuration.
            // Discovery is advisory: fail closed by sending nothing, never by
            // emitting an unsigned frame or throwing out of this thread.
            Log.e(TAG, "DROP discovery announce: required auth has no usable signing key")
            return
        }
        try {
            val out = DatagramPacket(reply, reply.size, packet.address, packet.port)
            sock.send(out) // unicast reply
            Log.i(TAG, "RX discover from=${result.parsed.from} ${packet.address.hostAddress}:${packet.port}; TX announce topology=${topology ?: "broker"} broker_hint=$brokerHint")
        } catch (_: Exception) {
        }
    }

    private fun verifyKeyFor(fromIdentity: String): ByteArray? = null

    internal fun makeAnnounce(): ByteArray? {
        val topo = topology
        val payload = jsonObj {
            put("device_id", deviceId)
            put("device_name", deviceName)
            put("ip", ip)
            put("broker_hint", brokerHint)
            // §14/§13/§17.3: only when this device declares a topology (it is the
            // coordinator). Omitted on the broker-client path → payload (and its
            // signature) stay byte-for-byte identical to the v1.2 announce.
            if (topo != null) {
                put("topology", topo)
                put("auth_mode", authMode.wire)
                put("key_mode", keyMode.wire)
            }
        }
        val explicitDeviceKey = deviceKey?.takeIf { it.isNotEmpty() }
        if (authMode == AuthMode.REQUIRED &&
            explicitDeviceKey == null && !Envelope.hasUsableKey(psk)) {
            return null
        }
        val env = if (keyMode == KeyMode.DERIVED && explicitDeviceKey != null) {
            // dk-only endpoint (§17.4): sign with our own device_key. This applies
            // to both broker-client and p2p coordinator announcements.
            Envelope.buildWithDeviceKey(explicitDeviceKey, authMode, "announce", "player:$deviceId", "all", payload)
        } else {
            // Auth-aware on every topology. In particular OPTIONAL + no key must
            // emit sig=""; calling the legacy always-signing build() here caused
            // SecretKeySpec(key.length == 0) and killed the entire API19 process
            // after transport_configure selected an open Broker.
            Envelope.buildWithMode(psk, authMode, "announce", "player:$deviceId", "all", payload, keyMode = keyMode)
        }
        return Envelope.toWire(env).toByteArray(Charsets.UTF_8)
    }

    fun updateIp(newIp: String) { ip = newIp }
    fun updateBrokerHint(hint: String) { brokerHint = hint }
    /** Keep announce.device_name in sync after §19 configure_device rename. */
    fun updateName(newName: String) { deviceName = newName }

    fun stop() {
        running = false
        try { socket?.close() } catch (_: Exception) {}
    }

    companion object {
        private const val TAG = "lmw.Discovery"
    }
}
